clients/lsp-kotlin.el

```diff
@@ -237,6 +237,33 @@ to Kotlin."
          (dolist (edit (-flatten selected-members))
            (lsp--apply-text-edits edit))))))
 
+(defun lsp-kotlin--get-filename (uri)
+  "Get the name of the buffer in a similar format to an import, calculating it based on URI."
+  (or (save-match-data
+        (when (string-match "kls:file://\\(.*\\)!/\\(.*\.\\(class\\|java\\|kt\\)\\)?.*" uri)
+          (format "%s"
+                  (replace-regexp-in-string "/" "." (match-string 2 url) t t))))
+      (error "Unable to match %s" uri)))
+
+(defun lsp-kotlin--resolve-uri (uri)
+  "Load a file corresponding to URI executing request to the kotlin server."
+  (let* ((buffer-name (lsp-kotlin--get-filename uri))
+         (file-location (concat (lsp-workspace-root) "/.cache/" buffer-name)))
```

Comment:
Should this be a constant value or dynamic based on the current project? 

It seems like if you work on multiple projects with different versions of the same libraries this caching could be misleading if all projects are sharing the same cache.

Reply:
Btw, it looks like clojure just uses a static path:

```
(defcustom lsp-clojure-workspace-dir (expand-file-name (locate-user-emacs-file "workspace/"))
  "LSP clojure workspace directory."
  :group 'lsp-clojure
  :risky t
  :type 'directory)

(defcustom lsp-clojure-workspace-cache-dir (expand-file-name ".cache/" lsp-clojure-workspace-dir)
  "LSP clojure workspace cache directory."
  :group 'lsp-clojure
  :risky t
  :type 'directory)
```

Reply:
In lsp-java it works like that: It obtains the "URL" which is version-specific and store it in <fileName>.metadata. The content is stored in <fileName> which is typically `package.ClassName.java`. Due to the fact that the file name might not be unique, we first check the metadata.

FWIW I am not sure if kotlin-ls is multiroot but for jdls you can have multiple versions of a particular class, thus having the cache workspace specific won't solve the issue with duplicates.

Reply:
ahh that makes sense! So the kls schema is something like `kls:file://<path_to_jar>!<package_path>`.

Would it make sense to store the file in `<path_to_cache>/<path_to_jar>.replace("/", ".")/<package_path>.replace("/", ".")`? This would always ensure file uniqueness. The only potential downside of this is storing multiple copies of the same version of the file when the gradle/mvn cache is cleared and files are re-downloaded with a different path.

Reply:
@aiguofer I guess that can work.

Reply:
great! just pushed those changes.

Does the same default as `cl-clojure` make sense or should we use a different default?